src/handle.rs

```diff
@@ -86,10 +86,18 @@ impl<Db: HasStorage> Handle<Db> {
             kind: EventKind::DidSetCancellationFlag,
         });
 
-        let mut clones = self.coordinate.clones.lock();
-        while *clones != 1 {
-            self.coordinate.cvar.wait(&mut clones);
+        {
+            let mut clones = self.coordinate.clones.lock();
+            while *clones != 1 {
+                self.coordinate.cvar.wait(&mut clones);
+            }
         }
+
+        // Force a new revision to clear the cancellation flag. 
+        // This is necessary to ensure non-mutable operations work after this.
+        let db =
+            Arc::get_mut(self.db_mut()).expect("other threads remain active despite cancellation");
+        db.runtime_mut().new_revision();
```

Comment:
I would prefer adding a `reset_cancellation_flag()` method to `db`. Creating a new revision comes at the cost that all queries need to be verified and avoiding the `get_mut` call in user-code when there are no "write" changes might not always be possible (or complicates the code a lot)